src/services/mapCode.ts

```diff
@@ -0,0 +1,484 @@
+            // That is, assuming we have the following code to start with:
+            //
+            // ```ts
+            // function foo() {
+            //     console.log("hello");
+            // }
+            // function bar() {
+            //     console.log("world");
+            // }
+            // function baz() {
+            //     console.log("!");
+            // }
+            // function other() {
+            //    console.log("Leave me alone!");
+            // }
+            // ```
+            //
+            // And we get a map request with the following code:
+            //
+            // ```ts
+            // function bar() {
+            //     console.log("world");
+            // }
+            // function baz() {
+            //     console.log("!");
+            // }
+            // function qux() {
+            //     console.log("?");
+            // }
+            // ```
+            //
+            // `bar` will be the beginning of our range, and we want to
+            // have `baz` in the original be the end of our replacement
+            // range. `other` should be left intact, while `qux` should be
+            // "injected" before it.
+            //
+            // We do this by scanning the scope for all nodes that have a
+            // match in the incoming map request code, then sorting and
+            // picking first and last by position, and that will be our
+            // range for replacement.
+            const start = (scope as Block).statements.find(stmt => changes.some(node => matchNode(node, stmt)));
+            const end = findLast((scope as Block).statements, stmt => changes.some(node => matchNode(node, stmt)));
+            // This should generally never be the case at this point, but just in case...
+            if (start && end) {
+                forEach(changes, wipeNode);
+                changeTracker.replaceNodeRangeWithNodes(
+                    originalFile,
+                    start,
+                    end,
+                    changes,
+                );
+                // はい、できました！
+                return;
+            }
+        }
+    }
+
+    // If we have no frame of reference, we'll just insert at the end of the
+    // nearest statements scope, or, if we can't find a block, at the top
+    // level of the source file.
+    let scopeStatements: NodeArray<Statement> = originalFile.statements;
+    for (const location of focusLocations) {
+        const block = findAncestor(
+            getTokenAtPosition(originalFile, location.start),
+            isBlock,
+        );
+        if (block) {
+            scopeStatements = block.statements;
+            break;
+        }
+    }
+    forEach(changes, wipeNode);
+    changeTracker.insertNodesAfter(
+        originalFile,
+        scopeStatements[scopeStatements.length - 1],
+        changes,
+    );
+}
+
+function matchNode(a: Node, b: Node): boolean {

```

Comment:
This is a lot more careful and precise than the existing diff-matching code, I think. Is it going to cause a lot more results to be insertions instead of diffs? Is that a good thing?

My first thoughts are: Yes and Yes. I think I'd prefer a duplicate thing to be inserted cleanly next to the old thing instead of a replacement overwriting it badly.

Reply:
No, and Yes: more precise matching increases the odds that we'll find a "reasonably accurate" match, and thus be able to replace existing code. The thing about this is that, a lot of times, it won't actually insert anything _next to_ where you intended: it'll shove it right in the middle of a definition, deeply nested and indented where it makes absolutely no sense, which is something we want to avoid.